README.md
# overflow-color [![Build Status][travis badge]][travis link] ![0 dependency][dependency badge] ![npm bundle size (minified + gzip)][size badge]

[![Demo](fixtures/demo.gif)][demo]

Try it on your smartphone or with a trackpad on MacOS: [git.io/overflow][demo] 
([https://dimitrinicolas.github.io/overflow-color/][demo])

A simple script that automatically switches CSS `html` `background-color` 
according to scroll position.

This package is on npm

```console
npm i overflow-color
```

## Usage

You simply must add the browser minified script 
[dist/overflow-color.umd.min.js](dist/overflow-color.umd.min.js) and the two 
attributes `data-oc-top` and `data-oc-bottom` to your body tag.

```html
<body data-oc-top="red" data-oc-bottom="blue">
    <!-- ... -->
    <script src="dist/overflow-color.umd.min.js" async></script>
</body>
```

You can use the shortcut `data-oc` by separating the two values with a comma.

```html
<body data-oc="red,blue">
```

### With a module bundler

You can import this package with a simple require or import.

```javascript
require('overflow-color');
// with ES6+
import 'overflow-color';
```

You don't have anything else to do, the script is automatically launched when 
the DOM content is loaded.

#### Manually re-check scroll position

When you update body's overflow color attribute or make a huge change to the DOM 
(e.g. changing page with Angular) and that the vertical scroll position stayed
at the top, the new document height may be smaller than the previous one, and
`overflow-color` should set the overflow color to the bottom one, but it has not
be called by scroll event. In this specific case, you can use the default
exported function `updateOverflowColor`:

```javascript
import updateOverflowColor from 'overflow-color';

/** Change DOM or body's overflow color attribute */

updateOverflowColor();
```

## CSS tricks, Behind the Scenes

This library will wrap all the body content inside a `<div data-oc-wrap>`.
Then it set to the wrapper the same background as the `body`, and set `body`'s 
background to `transparent`.

Add the `data-oc-outside` attribute to any body's immediate children that you
don't want to be included into this wrap element at DOM content load.

when the document is loaded:

```html
<!doctype html>
<head>
    <!-- ... -->
    <style>
        /* your style */
        body {
            background: lightgrey;
        }
    </style>
    <style>
        /* style added by the script */
        html { background: blue; }
    </style>
</head>
<body data-oc="red,blue" style="background: transparent; /* style added by the script */ ">
    <!-- wrapper added by the script -->
    <div data-oc-wrap style="background: lightgrey;">
        <!-- ... your body's childs -->
        <script src="dist/overflow-color.umd.min.js" async></script>
    </div>
</body>
```

You can read the [Designer News discussion][dn] about the different tried 
tricks.

## Browsers compatibility

I successfully tested this library, on:  
**Mac OS Mojave**: Safari v11, Google Chrome v70 and Opera v51  
**iOS 11**: Safari, Google Chrome v68, Firefox 12 and Microsoft Edge v42

Unfortunately, Firefox on MacOS and Opera Mini on iOS doesn't show the
over-scroll color even without this library.

This library pass an E2E test through Chrome with Cypress.

## Build

Compile with Rollup:

```console
npm run build
```

Build and test with Cypress:

```console
npm test
```

## License

This project is licensed under the [MIT license](LICENSE).

[travis badge]: https://travis-ci.org/dimitrinicolas/overflow-color.svg?branch=master
[travis link]: https://travis-ci.org/dimitrinicolas/overflow-color
[dependency badge]: https://img.shields.io/badge/dependencies-0-brightgreen.svg
[size badge]: https://img.shields.io/bundlephobia/minzip/overflow-color.svg

[demo]: https://dimitrinicolas.github.io/overflow-color/
[dn]: https://www.designernews.co/stories/91663-switch-html-background-color-in-order-to-get-two-overflow-scrolling-colors-on-smartphones


CHANGELOG.md
# Change Log
This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 2.3.0 - 2018-11-22
### Added
- A `data-oc-outside` attribute to exclude a body's immediate children to be put inside `data-oc-wrap` element

## 2.2.0 - 2018-11-22
### Added
- The default exported function is now `update` and check for attribute and scroll change

## 2.1.5 - 2018-10-25
### Changed
- Improved performance by passively listening to scroll event

## 2.1.4 - 2018-08-03
### Changed
- Improved performance by throttling the scroll event listener through requestAnimationFrame.
- Now default exporting a force `checkScroll` function

## 2.1.3 - 2018-05-14
### Changed
- Background color switch process for better performances.

## 2.1.2 - 2018-05-11
### Changed
- Now bundling the library with Rollup and exporting 3 versions: `main`, `module` and `browser` respectively as follows: `overflow-color.cjs.js`, `overflow-color.esm.js` and `overflow-color.umd.js` (plus `overflow-color.umd.min.js`).

## 2.1.1 - 2018-05-10
### Changed
- The test for page loaded or now, so the script can now be loaded `async`.

## 2.1.0 - 2018-03-10
### Added
- A shortcut `data-oc` for `data-oc-top` and `data-oc-bottom` combined.

## 2.0.1 - 2018-02-21
### Fixed
- A bug on Firefox.

## 2.0.0 - 2018-02-06
- Complete rework.

## 1.0.0 - 2017-03-02
- Initial release.

cypress/integration/test.js
// / <reference types="Cypress" />

context('Test', () => {
  beforeEach(() => {
    cy.viewport(300, 600).visit('http://localhost:8000');
  });

  it('page contain github link', () => {
    cy.get('.header a')
      .should('have.attr', 'href')
      .and('include', 'github');
  });

  it('overflow color wrap background is white', () => {
    cy.get('[data-oc-wrap]').should('have.css', 'background-color', 'rgb(255, 255, 255)');
  });

  it('html first background is red', () => {
    cy.get('html').should('have.css', 'background-color', 'rgb(255, 0, 0)');
  });

  it('html page end background is blue', () => {
    cy.get('p:last-of-type').scrollIntoView();
    cy.get('html').should('have.css', 'background-color', 'rgb(0, 0, 255)');
  });

  it('html page top background is blue', () => {
    cy.get('p:last-of-type').scrollIntoView();
    cy.scrollTo(0);
    cy.get('html').should('have.css', 'background-color', 'rgb(255, 0, 0)');
  });

  it('update body attribute', () => {
    cy.get('body').invoke('attr', 'data-oc', 'lime,yellow');
    cy.scrollTo(0);
    cy.window()
      .then(
        win => new Cypress.Promise(resolve => {
          win.updateOverflowColor();
          resolve();
        })
      )
      .then(() => {
        cy.get('html').should('have.css', 'background-color', 'rgb(0, 255, 0)');
        cy.get('p:last-of-type').scrollIntoView();
        cy.get('html').should('have.css', 'background-color', 'rgb(255, 255, 0)');
      });
  });

  it('outside element', () => {
    cy.get('body')
      .find('> [data-oc-outside]')
      .should('exist');
    cy.get('[data-oc-wrap]')
      .find('[data-oc-outside]')
      .should('not.exist');
    cy.get('[data-oc-wrap]')
      .find('+ [data-oc-outside]')
      .should('exist');
  });
});


cypress/integration/examples/test.js
// / <reference types="Cypress" />

context('Test', () => {
  beforeEach(() => {
    cy.viewport(300, 600).visit('http://localhost:8000');
  });

  it('page contain github link', () => {
    cy.get('.header a')
      .should('have.attr', 'href')
      .and('include', 'github');
  });

  it('overflow color wrap background is white', () => {
    cy.get('[data-oc-wrap]').should('have.css', 'background-color', 'rgb(255, 255, 255)');
  });

  it('html first background is red', () => {
    cy.get('html').should('have.css', 'background-color', 'rgb(255, 0, 0)');
  });

  it('html page end background is blue', () => {
    cy.get('p:last-of-type').scrollIntoView();
    cy.get('html').should('have.css', 'background-color', 'rgb(0, 0, 255)');
  });

  it('html page top background is blue', () => {
    cy.get('p:last-of-type').scrollIntoView();
    cy.scrollTo(0);
    cy.get('html').should('have.css', 'background-color', 'rgb(255, 0, 0)');
  });
});


src/main.js
let topColor;
let bottomColor;
let currentColor;

let styleTag;

const setBgColor = function(color) {

	if (currentColor !== color) {
		currentColor = color;

		let css = `html{background:${color};}`;

		if (!styleTag) {
			styleTag = document.createElement('style');
			let head = document.head || document.getElementsByTagName('head')[0];
			head.appendChild(styleTag);
		}

		if (styleTag.styleSheet) {
		  styleTag.styleSheet.cssText = css;
		}
		else {
			styleTag.innerHTML = css;
		}
	}

}

const checkScroll = function() {

  if (document.body.scrollHeight === window.innerHeight) {
    setBgColor(bottomColor);
  }
  else {
		let scrollFromMiddle = window.innerHeight - document.body.scrollHeight + 2 * Math.max(document.body.scrollTop, document.documentElement.scrollTop);
		setBgColor(scrollFromMiddle < 0 ? topColor : bottomColor);
  }

}

const initOverflowColor = function() {

  let shortcutEl = document.querySelector('[data-oc]');
  if (shortcutEl) {
    let split = shortcutEl.getAttribute('data-oc').split(',');
    if (split.length > 1) {
      topColor = split[0];
      bottomColor = split[1];
    }
    else if (split.length === 1) {
      topColor = bottomColor = split[0];
    }
  }
  else {
    let topEl = document.querySelector('[data-oc-top]');
    let bottomEl = document.querySelector('[data-oc-bottom]');
    if (topEl) {
      topColor = topEl.getAttribute('data-oc-top');
    }
    if (bottomEl) {
      bottomColor = bottomEl.getAttribute('data-oc-bottom');
    }
  }

  if (topColor || bottomColor) {

		if (!topColor && bottomColor) {
			topColor = bottomColor;
		}
		else if (topColor && !bottomColor) {
			bottomColor = topColor;
		}

    let bodyComputedStyle = window.getComputedStyle(document.body, null);
    let bodyBackground = bodyComputedStyle.getPropertyValue('background');
    if (bodyBackground === '' || (bodyComputedStyle.getPropertyValue('background-color') === 'rgba(0, 0, 0, 0)' && bodyBackground.substring(21, 17) === 'none')) {
      bodyBackground = 'white';
    }
    document.body.style.background = 'transparent';
    let bodyWrap = document.createElement('div');
    bodyWrap.setAttribute('data-oc-wrap', '');
    bodyWrap.style.background = bodyBackground;
    for (let i = 0, l = document.body.childNodes.length; i < l; i++) {
      bodyWrap.appendChild(document.body.childNodes[0]);
    }
    document.body.appendChild(bodyWrap);

    checkScroll();
    if (typeof window.addEventListener !== 'undefined') {
      window.addEventListener('scroll', checkScroll, false);
      window.addEventListener('resize', checkScroll, false);
    }
    else {
      window.attachEvent('scroll', checkScroll);
      window.attachEvent('resize', checkScroll);
    }

  }

}

if (['interactive', 'complete', 'loaded'].indexOf(document.readyState) >= 0) {
  initOverflowColor();
}
else if (typeof document.addEventListener !== 'undefined') {
  document.addEventListener('DOMContentLoaded', initOverflowColor, false);
}
else {
  document.attachEvent('onreadystatechange', initOverflowColor);
}

export default initOverflowColor;


src/index.js
const ATTRIBUTE_PREFIX = 'data-oc';

let topColor;
let bottomColor;

let currentBgColor;
let styleTag;

let lastScrollY;
let ticking = false;

/**
 * Request animation frame polyfill
 * @param {function} callback
 */
const requestAnimFrame = (() => {
  return (
    window.requestAnimationFrame
    || window.webkitRequestAnimationFrame
    || window.mozRequestAnimationFrame
    || (callback => {
      window.setTimeout(callback, 1000 / 60);
    })
  );
})();

/**
 * If needed, set the new new color as
 * html background
 * @param {string} color
 */
const setBgColor = color => {
  if (currentBgColor !== color) {
    currentBgColor = color;
    const css = `html { background: ${currentBgColor}; }`;

    if (!styleTag) {
      styleTag = document.createElement('style');
      const head = document.head || document.getElementsByTagName('head')[0];
      head.appendChild(styleTag);
    }

    if (styleTag.styleSheet) {
      styleTag.styleSheet.cssText = css;
    } else {
      styleTag.innerHTML = css;
    }
  }
};

/**
 * Checks the scroll position and determines
 * the overflow color to set between the
 * topColor and the bottomColor
 */
const checkScroll = () => {
  lastScrollY = window.scrollY;
  if (!ticking && (topColor || bottomColor)) {
    requestAnimFrame(() => {
      const scrollHeight = document.body.scrollHeight;
      const innerHeight = window.innerHeight;
      if (scrollHeight === innerHeight) {
        setBgColor(bottomColor);
      } else {
        setBgColor(innerHeight - scrollHeight + 2 * lastScrollY < 0 ? topColor : bottomColor);
      }
      ticking = false;
    });
    ticking = true;
  }
};

/**
 * Update colors and check scroll
 */
const updateOverflowColor = () => {
  topColor = null;
  bottomColor = null;

  const shortcutAttributeEl = document.querySelector(`[${ATTRIBUTE_PREFIX}]`);
  if (shortcutAttributeEl) {
    const split = shortcutAttributeEl.getAttribute(ATTRIBUTE_PREFIX).split(',');
    if (split.length > 1) {
      topColor = split[0];
      bottomColor = split[1];
    } else if (split.length === 1) {
      topColor = bottomColor = split[0];
    }
  } else {
    const topAttributeEl = document.querySelector(`[${ATTRIBUTE_PREFIX}-top]`);
    const bottomAttributeEl = document.querySelector(`[${ATTRIBUTE_PREFIX}-bottom]`);
    if (topAttributeEl) {
      topColor = topAttributeEl.getAttribute(`${ATTRIBUTE_PREFIX}-top`);
    }
    if (bottomAttributeEl) {
      bottomColor = bottomAttributeEl.getAttribute(`${ATTRIBUTE_PREFIX}-bottom`);
    }
  }

  if (!topColor && bottomColor) {
    topColor = bottomColor;
  } else if (topColor && !bottomColor) {
    bottomColor = topColor;
  }

  const bodyComputedStyle = window.getComputedStyle(document.body, null);
  let bodyComputedBackground = bodyComputedStyle.getPropertyValue('background');
  if (
    bodyComputedBackground === ''
    || (bodyComputedStyle.getPropertyValue('background-color') === 'rgba(0, 0, 0, 0)' && bodyComputedBackground.substring(21, 17) === 'none')
  ) {
    bodyComputedBackground = 'white';
  }
  document.body.style.background = 'transparent';

  checkScroll();
};

/**
 * Gets the two overflow colors and init the
 * window scroll and resize event listeners
 */
const initOverflowColor = () => {
  const bodyComputedStyle = window.getComputedStyle(document.body, null);
  let bodyComputedBackground = bodyComputedStyle.getPropertyValue('background');
  if (
    bodyComputedBackground === ''
    || (bodyComputedStyle.getPropertyValue('background-color') === 'rgba(0, 0, 0, 0)' && bodyComputedBackground.substring(21, 17) === 'none')
  ) {
    bodyComputedBackground = 'white';
  }
  document.body.style.background = 'transparent';

  const bodyWrapperEl = document.createElement('div');
  bodyWrapperEl.setAttribute(`${ATTRIBUTE_PREFIX}-wrap`, '');
  bodyWrapperEl.style.background = bodyComputedBackground;
  for (let i = document.body.childNodes.length - 1; i > 0; i--) {
    const child = document.body.childNodes[i];
    if (typeof child.getAttribute !== 'function' || child.getAttribute(`${ATTRIBUTE_PREFIX}-outside`) === null) {
      bodyWrapperEl.insertBefore(child, bodyWrapperEl.childNodes[0]);
    }
  }
  if (document.body.childNodes.length) {
    document.body.insertBefore(bodyWrapperEl, document.body.childNodes[0]);
  } else {
    document.body.appendChild(bodyWrapperEl);
  }

  updateOverflowColor();

  if (typeof window.addEventListener !== 'undefined') {
    window.addEventListener('scroll', checkScroll, { passive: true });
    window.addEventListener('resize', checkScroll, { passive: true });
  } else {
    window.attachEvent('scroll', checkScroll);
    window.attachEvent('resize', checkScroll);
  }
};

if (['interactive', 'complete', 'loaded'].indexOf(document.readyState) !== -1) {
  initOverflowColor();
} else if (typeof document.addEventListener !== 'undefined') {
  document.addEventListener('DOMContentLoaded', initOverflowColor, false);
} else {
  document.attachEvent('onreadystatechange', initOverflowColor);
}

window.updateOverflowColor = updateOverflowColor;

export default updateOverflowColor;


rollup.config.js
import babel from 'rollup-plugin-babel';
import resolve from 'rollup-plugin-node-resolve';
import commonjs from 'rollup-plugin-commonjs';
import pkg from './package.json';

export default [
	{
		input: 'src/index.js',
		output: {
			name: 'overflowColor',
			file: pkg.browser,
			format: 'umd'
		},
		plugins: [
      babel({
        presets: [
          [
            'env',
            { modules: false }
          ]
        ],
        plugins: [
          'external-helpers'
        ],
        externalHelpers: true
      }),
			resolve(),
			commonjs()
		]
	},
	{
		input: 'src/index.js',
		external: ['ms'],
		output: [
			{ file: pkg.main, format: 'cjs' },
			{ file: pkg.module, format: 'es' }
		],
    plugins: [
      babel({
        presets: [
          [
            'env',
            { modules: false }
          ]
        ],
        plugins: [
          'external-helpers'
        ],
        externalHelpers: true
      })
    ]
	}
];


overflow-color.js
var ocTopColor,
    ocBottomColor;

var ocStyleTag;

var currentColor;

var bodyBackground;

var bodyWrap;

function setBgColor(color) {

    currentColor = color;

    if (ocStyleTag) {
        ocStyleTag.parentNode.removeChild(ocStyleTag);
    }

    var css = 'html { background: ' + color + '; }',
        head = document.head || document.getElementsByTagName('head')[0];

    ocStyleTag = document.createElement('style');

    if (ocStyleTag.styleSheet){
        ocStyleTag.styleSheet.cssText = css;
    }
    else {
        ocStyleTag.appendChild(document.createTextNode(css));
    }

    head.appendChild(ocStyleTag);

}

function handleScroll() {

    if (document.body.scrollHeight === window.innerHeight) {
        setBgColor(ocBottomColor);
    }
    else {

        var diffTop = Math.max(document.body.scrollTop, document.documentElement.scrollTop);
        var diffBottom = document.body.scrollHeight - (diffTop + window.innerHeight);

        if (diffTop < diffBottom && currentColor !== ocTopColor) {
            setBgColor(ocTopColor);
        }
        else if (diffTop > diffBottom && currentColor !== ocBottomColor) {
            setBgColor(ocBottomColor);
        }

    }

}

function initOverflowColor() {

    var topEl = document.querySelector('[data-oc-top]'),
        bottomEl = document.querySelector('[data-oc-bottom]'),
        shortcutEl = document.querySelector('[data-oc]');

    if (shortcutEl) {
        var split = shortcutEl.getAttribute('data-oc').split(',');
        if (split.length > 1) {
            ocTopColor = split[0];
            ocBottomColor = split[1];
        }
        else if (split.length === 1) {
            ocTopColor = ocBottomColor = split[0];
        }
    }
    else {
        if (topEl) {
            ocTopColor = topEl.getAttribute('data-oc-top');
        }
        if (bottomEl) {
            ocBottomColor = bottomEl.getAttribute('data-oc-bottom');
        }
    }

    if (ocTopColor || ocBottomColor) {

        var bodyComputedStyle = window.getComputedStyle(document.body, null);
        bodyBackground = bodyComputedStyle.getPropertyValue('background');
        if (bodyBackground === '' || (bodyComputedStyle.getPropertyValue('background-color') === 'rgba(0, 0, 0, 0)' && bodyBackground.substring(21, 17) === 'none')) {
            bodyBackground = 'white';
        }
        document.body.style.background = 'transparent';
        bodyWrap = document.createElement('div');
        bodyWrap.setAttribute('data-oc-wrap', '');
        bodyWrap.style.background = bodyBackground;
        for (var i = 0, l = document.body.childNodes.length; i < l; i++) {
            bodyWrap.appendChild(document.body.childNodes[0]);
        }
        document.body.appendChild(bodyWrap);

        if (!ocTopColor && ocBottomColor) {
            ocTopColor = ocBottomColor;
        }
        else if (ocTopColor && !ocBottomColor) {
            ocBottomColor = ocTopColor;
        }

        if (document.body.scrollHeight > window.innerHeight) {
            setBgColor(ocTopColor);
        }
        else {
            setBgColor(ocBottomColor);
        }

        if (typeof window.addEventListener !== 'undefined') {
            window.addEventListener('scroll', handleScroll, false);
            window.addEventListener('resize', handleScroll, false);
        }
        else {
            window.attachEvent('scroll', handleScroll);
            window.attachEvent('resize', handleScroll);
        }

    }

}

if (['interactive', 'complete', 'loaded'].indexOf(document.readyState) >= 0) {
    initOverflowColor();
}
else if (typeof document.addEventListener !== 'undefined') {
    document.addEventListener('DOMContentLoaded', initOverflowColor, false);
}
else {
    document.attachEvent('onreadystatechange', initOverflowColor);
}
